### DISCUSSION TITLE
Audio is choppy/distorted under x86_64 Crostini/ChromeOS Linux Development Environment

### DISCUSSION THREAD

**Discussion by kurtbahartr** [ppy/osu#28366] [Q&A] [answered]
### Type

Game behaviour

### Bug description

The title should be obvious but the audio is choppy/distorted in game. I made sure it's specific to osu!lazer by testing the sounds on the Discord app I had installed on that very same container as that would've been the fastest way for me to test audio output on Crostini.

### Screenshots or videos

https://photos.app.goo.gl/Jo5wFPq89He3bAdP7

I had to use Google Photos because it's more than 10 MB in size and 10 MB is the maximum size GitHub allows when uploading videos to issues.

### Version

2024.521.2

### Logs

[compressed-logs.zip](https://github.com/user-attachments/files/15515656/compressed-logs.zip)

**Comment by bdach** [comment]
We don't really support such setups sorry. You're going to have to debug on your own (or find help from someone else that is running this same type of setup).

**Reply by kurtbahartr** [reply]
That's fine, I'll try to debug it and share the outcome and steps to actually fix it as a reply here.

**Reply by kurtbahartr** [reply]
Sigh I've completely forgotten to get back to you and I'm really sorry for that.

Here's the solution I've found to this problem of mine with the help of sineplusx, just remembered to send it as a reply to you just in case someone else happens to have the same type of setup with the same type of problem and ends up opening a duplicate of this: https://github.com/ppy/osu/discussions/28366#discussioncomment-9624201

**Comment by sineplusx** [comment]
This was already reported, see #25802

**Reply by kurtbahartr** [reply]
They've suggested trying out the Flatpak or Android version. Flatpak version is worth a shot, however the Android version doesn't run that well (lots of latency in input is the most prominent one). I'll be posting the outcome of the Flatpak version here.

EDIT: I tried to install the Flatpak version just now but it's marked as EOL apparently.
```
Info: app sh.ppy.osulazer branch stable is end-of-life, with reason:
   This application is no longer maintained.
error: The app sh.ppy.osulazer/x86_64/stable requires non-existent runtime environment org.gnome.Platform/i386/3.32
```

**Comment by sineplusx** [comment]
I am not familiar with Linux on Chromebooks, but this issue probably relates to the sound server and its configuration.
It would be interesting to know what distribution is used here (default is Debian I believe) and what sound server is in use (PulseAudio or PipeWire)

**Reply by kurtbahartr** [reply]
Yep, the distro that's installed on Chromebooks by default is Debian (currently Debian Bullseye) and the audio server in use is PipeWire. I've tried playing with some of the configurations shipped in Crostini (such as decreasing the sample rate from `48000`  to `44100` for the output) but to no avail. What's even weirder is this happens only on osu!lazer (at least on the AppImage version) and nowhere else. :/

**Reply by sineplusx** [reply]
The issue is a bit complex. It sounds like buffer underruns (choppy audio), but then the audio also speeds up and slows down randomly. The difference to just playing a file from disk with `ffmpeg` is the larger buffer size / latency which likely conceals the issue in your example.

Here are a few other things you could try :
- Check realtime process priority of PipeWire with a tool like `htop`
- Adapt the `quantum` value which controls buffer size on the sound server level. Can be checked with `pw-top`.
```
pw-metadata -n settings 0 clock.force-rate 44100
pw-metadata -n settings 0 clock.force-quantum 512
```
- Try setting custom buffer sizes for osu! itself, via environment variables
Note that the `quantum` values are samples, in the second command it is bytes. The buffer size has to be a multiple of the period size and at least 2x as large.
```
export PIPEWIRE_LATENCY=512/44100
export PIPEWIRE_ALSA='{ alsa.format=S32_LE alsa.channels=2 alsa.buffer-bytes=2048 alsa.period-bytes=1024 }'
```
- Try running the game in single-threaded mode

Also, can you make a screen recording of the game with Ctrl+F11 pressed twice? (frame graphs shown)

**Reply by kurtbahartr** [reply, marked as answer]
Pipe was running with a priority of `20` which I think is normal. As for the `pw-metadata` commands, those turned out to be exactly what I needed *except* I had to use the `quantum` value of `256` to get it just right. In the following link is me booting up osu! with just the `pw-metadata` commands run and playing the map for Flaklypa by Rameses B (sorry for horrible playthrough lmao I'm sleepy rn), also with frame graphs shown as you've requested.

https://photos.app.goo.gl/kBbeGjsPngL6oxku7

EDIT: I've also found the way to make these changes permanent;
1. Open up the file `/etc/pipewire/pipewire.conf.d/crostini.conf` - If it's not present, you might want to install `cros-pipe-config` first.
2. Then find the configs `default.clock.min-quantum` under `context.properties` and `audio.rate` under the audio devices listed in the configuration (They're "PCM Source" and "PCM Sink" as of the date of writing this).
3. Now set the values to `256` (for `min-quantum`) and `44100` (for `rate`) respectively.
4. After saving the changes to this file, restart the PipeWire service: `systemctl --user restart pipewire`
5. Now osu! should no longer have problems with audio!

**Reply by sineplusx** [reply]
Nice that it worked out. The screen recording I asked for was just to check if there is anything funny going on with the frametimes, but everything seems to be working fine now.

**Reply by kurtbahartr** [reply]
Yup, I've also made sure to post a heads-up about this to Google's official Chromebook support community so peeps at Google might notice this issue and get it sorted out in an update (since they're the ones developing the `cros-pipe-config` package from what I know).

**Comment by lewisdonofrio** [comment]
this works on my cromebook but if like replying to this it gets choppy as I type, how do fix this?  sounds fine if I don't do anything else?

**Reply by lewisdonofrio** [reply]
https://paste.debian.net/1324908/

**Reply by kurtbahartr** [reply]
1. Can you send a video recording for better demonstration and steps to reproduce?
2. Your paste has expired already.
